050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,832 @@
+But they will have different roles - they will be either controllers, or brokers, or will combine both of them.
+The controller role means the node is responsible for maintaining the cluster quorum, for the elections, discovery and for handling the cluster metadata instead of ZooKeeper.
+The broker role means the node is responsible for handling messages as a regular broker in the old ZooKeeper based clusters.
+And finally, the combined role means the node is responsible for both at the same time.
+
+The KRaft based Kafka clusters will be typically used in 3 different architectures:
+
+1. One or more nodes combining the controller and broker roles
+2. One or more nodes with the controller roles and one or more nodes with the broker roles
+3. One or more nodes with the combined role and one or more nodes with the broker roles
+
+In Strimzi, we want to support all of these architectures.
+But we also want to make it possible to transition between them.
+Since Apache Kafka plans to remove ZooKeeper support in its 4.0 release, supporting KRaft mode in its different form is not really an option for us but something we have to do.
+
+### Stretch clusters
+
+There seems to be increased demand Kafka clusters stretched across multiple Kubernetes clusters.
+Apache Kafka itself has latency limitations and might not run well when stretched over distant clusters with big latency.
+But there are many use cases where stretched clusters might be useful.
+For example:
+
+* When running closely colocated Kubernetes clusters
+* When migrating from old Kubernetes cluster to new Kubernetes cluster
+
+## Proposal
+
+This proposal suggests introduction of Kafka _node pools_.
+_Node pools_ are groups of Kafka nodes which will have the same configuration.
+A Kafka cluster can consist of one or more pools.
+Each node pool can belong to only one Kafka cluster.
+
+### `KafkaNodePool` custom resource
+
+A new custom resource called `KafkaNodePool` will be added to Strimzi.
+One `KafkaNodePool` resource will represent one _node pool_.
+It will take over some of the configuration from the `Kafka` CR, reuse other parts from it and add some new configuration options as well.
+The `KafkaNodePool` resource will be a _public_ resource.
+It will be created and managed directly by users (unlike for example the `StrimziPodSet` resource which is used only internally by Strimzi and should not be touched by the users).
+
+The _cluster-wide_ configuration options will remain part of the `Kafka` CR.
+While the _pool-specific_ options will be taken from `KafkaNodePool` resources
+
+For the initial implementation, the following options are proposed for the `KafkaNodePool` resource and its `.spec` section:
+
+* `replicas` indicating the number of nodes in the pool
+* `storage` indicating the storage of the nodes in this pool
+* `resources` indicating the amount of resources (requests / limits) assigned to the nodes in this pool
```

Comment:
I did not consider this. From my experience, the majority of Strimzi users do not run large fleets of uniform clusters. They tend to either have a limited number of centrally managed clusters or many different clusters with different configurations. So I'm not sure this factory approach fits - I would be afraid it increases the complexity for both users as well as on the code level without too much added value.

You can probably also combine other tools such as GitOps to create a factory-style approach like this - for example with a Helm Chart which limits the configurability of some parts to decrease the complexity / increase the uniformity

Reply:
Yeah my background is definitely showing through :D 

Defaulting to the factory model probably would be a small increment in complexity for users managing small numbers of clusters and potentially pointing them towards a different pattern, which is probably more opinionated than the project want to be. I do wonder if we can find a middle ground, this ties in with [my thinking](https://github.com/strimzi/proposals/pull/77#discussion_r1162112642) around `stroagePools`, and support either an inline definition or a pointer to a `spec`. The default option would be an inline definition. 

To turn from the user complexity to the implementation complexity for a moment, again I don't actually think it's that large, or rather doesn't need to be. In my mental model the reconciliation would effectively split into two phases. An assembly phase stitching everything together (i.e. following the pointers to the `spec` definitions so as to appear as if inline) which would output a fully rendered model to the second phase of the reconciler which would then operate on that rendered model and thus wouldn't care if it contained pointers or inline definitions were used. 

I guess that would lead to questions about validation of the separate `spec` resources. I think, initially at least, they could just be handled like the inline version would be in the current proposal however there would be the option to validate them in a dedicated reconciler.

Reply:
Sorry, I'm getting a bit lost and I'm still not sure I understand what would be the actual benefit.

The storage configuration is inlined already today and as part of this proposal. I also don't follow what you mean by the _factory model_. There is no factory here. The resources belong to a single unique cluster in a single namespace. There is no reuse. Especially with storage, which is mostly configurable at creation only, it would be quite hard to share the configuration and the changes would very quickly get completely out of sync without any real possibility to fix it.

For example - imagine someone creates the storage configuration for `gp2` volumes and uses it to create cluster A. And then changes the _shared_ storage configuration to `gp3` and creates cluster B. Suddenly, you have a single storage configuration that is always invalid for either cluster A or cluster B.

Reply:
> While I don't have the data to back up the hunch I'd guess people operating diverse sets of Kafka clusters have very similar Zookeeper clusters (Zookeeper having very different scaling needs to Kafka and I suspect the same holds for Kraft controllers) and would thus benefit from having a standard definition of a controller cluster while maintaining the flexibility to run very different scales of broker cluster.

There will be probably similar pieces of configuration. But I think they might differ and different users might want different pieces to be _reusable_. You talked about storage and resources. Others would want to reuse affinity / tolerations / topology spread constraints rules, etc. I don't think you can solve this by splitting the Strimzi configuration into many small pieces. I think the solution is to solve it on the upper levels.

So, if you are running a large and centrally managed fleet of Kafka clusters, you will probably rely on other means anyway. For example on a custom operator for managing the clusters. Or on some GItOps flows and custom Helm Charts which allows you to decide which configuration should be shared and which should be up to customization. I think leaving the unification of some parts of the custom resources to this layer is much better then creating a separate resource for every single field someone wants share for their fleet of Kafka clusters.